File: fetch_financial_data.py
#!/usr/bin/env python3
"""
Fetch financial data from multiple sources and save to local JSON + formatted sheets.
Usage:
    python3 fetch_financial_data.py --ticker VIST --period quarterly
    python3 fetch_financial_data.py --ticker AAPL --period annual --fmp-key YOUR_KEY
"""

import argparse
import json
import logging
import os
import sys
from pathlib import Path

import pandas as pd

from src.services.financial_data_service import FinancialDataService
from src.utils.report_formatter import ReportFormatter

logging.basicConfig(
    level=logging.WARNING,
    format="%(levelname)s - %(message)s",
)
logger = logging.getLogger(__name__)


def main():
    parser = argparse.ArgumentParser(
        description="Fetch financial data from multiple sources and save locally"
    )

    parser.add_argument(
        "--ticker",
        required=True,
        help="Stock ticker symbol (e.g., VIST, AAPL)",
    )

    parser.add_argument(
        "--period",
        choices=["annual", "quarterly"],
        default="quarterly",
        help="Report period (default: quarterly)",
    )

    parser.add_argument(
        "--fmp-key",
        default=os.getenv("FMP_API_KEY"),
        help="FMP API key for peer data (env: FMP_API_KEY)",
    )

    parser.add_argument(
        "--output-dir",
        default="./data/financial_reports",
        help="Directory to save JSON output (default: ./data/financial_reports)",
    )

    args = parser.parse_args()

    try:
        print(f"Fetching {args.period} data for {args.ticker.upper()}...")

        # Fetch data from multiple sources
        data_service = FinancialDataService(fmp_api_key=args.fmp_key)
        financial_data = data_service.get_company_data(args.ticker, period=args.period)

        # Prepare data for JSON export
        data_to_save = {
            "ticker": financial_data["ticker"],
            "period": financial_data["period"],
            "sources_used": financial_data.get("sources", {}),
        }

        # Add overview
        if financial_data.get("overview"):
            data_to_save["overview"] = financial_data.get("overview")

        # Add income statement
        income_stmt = financial_data.get("income_statement")
        if isinstance(income_stmt, pd.DataFrame) and not income_stmt.empty:
            data_to_save["income_statement"] = income_stmt.to_dict()

        # Add balance sheet
        balance_sheet = financial_data.get("balance_sheet")
        if isinstance(balance_sheet, pd.DataFrame) and not balance_sheet.empty:
            data_to_save["balance_sheet"] = balance_sheet.to_dict()

        # Add cash flow
        cash_flow = financial_data.get("cash_flow")
        if isinstance(cash_flow, pd.DataFrame) and not cash_flow.empty:
            data_to_save["cash_flow"] = cash_flow.to_dict()

        # Add ratios
        ratios = financial_data.get("ratios")
        if isinstance(ratios, pd.DataFrame) and not ratios.empty:
            data_to_save["ratios"] = ratios.to_dict()

        # Add peers
        peers = financial_data.get("peers")
        if peers:
            data_to_save["peers"] = peers

        # Save to JSON
        output_dir = Path(args.output_dir)
        output_dir.mkdir(parents=True, exist_ok=True)

        output_file = output_dir / f"{args.ticker.lower()}_{args.period}.json"

        with open(output_file, "w") as f:
            json.dump(data_to_save, f, indent=2, default=str)

        print(f"JSON to {output_file.relative_to(Path.cwd())}")

        # Format data for sheets
        formatter = ReportFormatter(financial_data)
        sheets = formatter.generate_all_sheets()

        # Save formatted sheets as CSV for inspection
        sheets_dir = output_dir / "sheets"
        sheets_dir.mkdir(parents=True, exist_ok=True)

        sheet_count = 0
        for sheet_name, df in sheets.items():
            if df is not None and not df.empty:
                sheet_file = (
                    sheets_dir
                    / f"{args.ticker.lower()}_{sheet_name.lower().replace(' ', '_')}.csv"
                )
                df.to_csv(sheet_file, index=False)
                sheet_count += 1

        print(
            f"Generated {sheet_count} sheets in {sheets_dir.relative_to(Path.cwd())}/"
        )
        print(
            f"\nData sources: {', '.join(s for s in financial_data.get('sources', {}).values() if s)}"
        )

        return 0

    except KeyboardInterrupt:
        print("\n✗ Cancelled")
        return 130

    except Exception as e:
        print(f"✗ Error: {e}")
        logger.debug(e, exc_info=True)
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: ts/cli/fetchFinancialData.ts
#!/usr/bin/env node

import { program } from "commander";
import { config } from "dotenv";
import { FinancialDataOrchestrator } from "../services/financialDataOrchestrator.js";
import { ExcelWriter } from "../gateway/excelWriter.js";
import { CsvWriter } from "../utils/csvWriter.js";
import * as fs from "fs";
import * as path from "path";

config();

interface Options {
  ticker: string;
  period?: "annual" | "quarterly";
  output?: string;
  format?: "json" | "csv" | "excel";
}

async function fetchFinancialData(options: Options): Promise<void> {
  console.log(`\nFetching ${options.period} financial data for ${options.ticker}...\n`);

  const orchestrator = new FinancialDataOrchestrator(process.env["FMP_API_KEY"]);

  const result = await orchestrator.getCompanyData(
    options.ticker.toUpperCase(),
    options.period ?? "quarterly"
  );

  console.log("Data retrieved successfully!");
  console.log(`\nCompany: ${result.ticker}`);
  console.log(`Period: ${result.period}`);

  if (result.overview) {
    console.log(`\nOverview:`);
    console.log(`  Currency: ${result.overview.currency}`);
    console.log(`  Market Cap: ${result.overview.marketCap}`);
    console.log(`  PE Ratio: ${result.overview.trailingPE}`);
    console.log(`  Dividend Yield: ${result.overview.dividendYield}%`);
  }

  if (result.incomeStatement) {
    const dates = Object.keys(result.incomeStatement);
    console.log(`\nIncome Statements: ${dates.length} periods found`);
  }

  if (result.balanceSheet) {
    const dates = Object.keys(result.balanceSheet);
    console.log(`Balance Sheets: ${dates.length} periods found`);
  }

  if (result.cashFlow) {
    const dates = Object.keys(result.cashFlow);
    console.log(`Cash Flow Statements: ${dates.length} periods found`);
  }

  const outputDir = options.output || "./data/financial_data";
  fs.mkdirSync(outputDir, { recursive: true });

  const format = options.format ?? "json";

  if (format === "json" || format === "csv" || format === "excel") {
    const filename = `${result.ticker}_${result.period}`;
    const fileExt = format === "json" ? "json" : format === "csv" ? "csv" : "xlsx";
    const filePath = path.join(outputDir, `${filename}.${fileExt}`);

    if (format === "json") {
      const jsonData = JSON.stringify(result, null, 2);
      fs.writeFileSync(filePath, jsonData);
      console.log(`\n✅ JSON saved to: ${filePath}`);
    } else if (format === "csv") {
      const csvWriter = new CsvWriter();
      const data: unknown[][] = [];

      data.push(["Financial Data Export"]);
      data.push(["Ticker", result.ticker]);
      data.push(["Period", result.period]);
      data.push([]);

      if (result.overview) {
        data.push(["Overview Data"]);
        data.push(["Currency", String(result.overview.currency)]);
        data.push(["Market Cap", String(result.overview.marketCap)]);
        data.push(["PE Ratio", String(result.overview.trailingPE)]);
        data.push([]);
      }

      if (result.incomeStatement) {
        data.push(["Income Statement"]);
        const dates = Object.keys(result.incomeStatement);
        if (dates.length > 0) {
          const firstDate = dates[0];
          if (firstDate !== undefined) {
            const firstStmt = result.incomeStatement[firstDate];
            if (firstStmt) {
              const keys = Object.keys(firstStmt);
              data.push(["Date", ...keys]);
              for (const date of dates) {
                const stmt = result.incomeStatement[date];
                if (stmt) {
                  data.push([date, ...keys.map((k) => String(stmt[k] ?? ""))]);
              }
            }
          }
        }
      }

      csvWriter.writeFile(filePath, data);
      console.log(`\n✅ CSV saved to: ${filePath}`);
    } else if (format === "excel") {
      const excelWriter = new ExcelWriter();
      const sheets = [];

      const overviewData: unknown[][] = [["Overview Data"]];
      if (result.overview) {
        overviewData.push(["Currency", result.overview.currency]);
        overviewData.push(["Market Cap", result.overview.marketCap]);
        overviewData.push(["PE Ratio", result.overview.trailingPE]);
        overviewData.push(["Dividend Yield", result.overview.dividendYield]);
      }
      sheets.push({ name: "Overview", data: overviewData });

      if (result.incomeStatement) {
        const incomeData: unknown[][] = [["Income Statement"]];
        const dates = Object.keys(result.incomeStatement);
        if (dates.length > 0) {
          const firstStmt = result.incomeStatement[dates[0]];
          if (firstStmt) {
            const keys = Object.keys(firstStmt);
            incomeData.push(["Date", ...keys]);
            for (const date of dates) {
              const stmt = result.incomeStatement[date];
              if (stmt) {
                incomeData.push([date, ...keys.map((k) => stmt[k] ?? "")]);
              }
            }
          }
        }
        sheets.push({ name: "Income", data: incomeData });
      }

      if (result.balanceSheet) {
        const balanceData: unknown[][] = [["Balance Sheet"]];
        const dates = Object.keys(result.balanceSheet);
        if (dates.length > 0) {
          const firstDate = dates[0];
          if (firstDate !== undefined) {
            const firstStmt = result.balanceSheet[firstDate];
            if (firstStmt) {
              const keys = Object.keys(firstStmt);
              balanceData.push(["Date", ...keys]);
              for (const date of dates) {
                const stmt = result.balanceSheet[date];
                if (stmt) {
                  balanceData.push([date, ...keys.map((k) => stmt[k] ?? "")]);
              }
            }
          }
        }
        sheets.push({ name: "Balance", data: balanceData });
      }

      excelWriter.writeFile(filePath, sheets);
      console.log(`\n✅ Excel saved to: ${filePath}`);
    }
  }

  console.log(`\nData sources:`);
  console.log(`  Overview: ${result.sources.overview || "N/A"}`);
  console.log(`  Income Statement: ${result.sources.incomeStatement || "N/A"}`);
  console.log(`  Balance Sheet: ${result.sources.balanceSheet || "N/A"}`);
  console.log(`  Cash Flow: ${result.sources.cashFlow || "N/A"}`);
  console.log(`  Peers: ${result.sources.peers || "N/A"}`);
  console.log("");
}

program
  .name("financial-data")
  .description("Fetch and export financial data from Yahoo Finance")
  .version("1.0.0");

program
  .command("fetch <ticker>")
  .description("Fetch financial data for a company")
  .option("-p, --period <period>", "Period (annual or quarterly)", "quarterly")
  .option("-o, --output <path>", "Output directory", "./data/financial_data")
  .option("-f, --format <format>", "Output format (json, csv, excel)", "json")
  .action(async (ticker: string, options: Record<string, string>) => {
    try {
      await fetchFinancialData({
        ticker,
        period: (options["period"] as "annual" | "quarterly") ?? "quarterly",
        output: options["output"],
        format: (options["format"] as "json" | "csv" | "excel") ?? "json",
      });
    } catch (error) {
      console.error("Error:", error instanceof Error ? error.message : error);
      process.exit(1);
    }
  });

program.parse();
